### PR TITLE
[Taxonomy] Allow parent locale for taxonSlugGenerator

### LINKED ISSUE #8213: [Taxonomy] TaxonSlugGenerator allow parent locale
When generating a slug for the taxon from the parent you cannot set desired locale.
Ex:
* code: footwear
  * translation slugs:
    * en: footwear
    * ro: incaltaminte
* code: sneakers
  * translation slugs:
    * en: footwear/sneakers
    * ro: **footwear/adidasi**

it should be: **incaltaminte/adidasi**

### PR DESCRIPTION
**Pull request by theodormanolescu**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | no |
| Related tickets | fixes #8213 |
| License         | MIT |

### CODE REVIEW

File: src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php [hunk 1]
```diff
@@ -52,6 +52,10 @@ public function generate($name, $parentId = null)
         $parent = $this->taxonRepository->find($parentId);
         Assert::notNull($parent, sprintf('There is no parent taxon with id %d.', $parentId));
 
+        if ($parentLocale) {
+            $parent->setCurrentLocale($parentLocale);
+        }
+
         return $parent->getSlug().self::SLUG_SEPARATOR.$taxonSlug;
```
Comment (lchrusciel) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php]: `$parent->getTranslation($locale)->getSlug().self::SLUG_SEPARATOR.$taxonSlug` will be a better solution. `$parent->getTranslation($locale)` could be even extracted to variable.

File: src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php [hunk 2]
```diff
@@ -39,7 +39,7 @@ public function __construct(TaxonRepositoryInterface $taxonRepository)
     /**
      * {@inheritdoc}
      */
-    public function generate($name, $parentId = null)
+    public function generate($name, $parentId = null, $parentLocale = null)
```
Comment (lchrusciel) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php]: `parentLocale` -> `locale`

**Comment by lchrusciel** [comment]
The tests fail because there is outdated spec. Can you also adjust TaxonSlugGeneratorSpec?

**Comment by theodormanolescu** [comment]
Sure, will update today

**Comment by lchrusciel** [comment]
Thanks! :)

**Comment by lchrusciel** [comment]
Can you also rebase to the master?

**Comment by theodormanolescu** [comment]
Hey, i currently don't have internet access until Thursday will fix it then if there is no rush. :)

**Comment by lchrusciel** [comment]
We will wait :)

File: src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php [hunk 3]
```diff
@@ -52,6 +52,6 @@ public function generate($name, $parentId = null)
         $parent = $this->taxonRepository->find($parentId);
         Assert::notNull($parent, sprintf('There is no parent taxon with id %d.', $parentId));
 
-        return $parent->getSlug().self::SLUG_SEPARATOR.$taxonSlug;
+        return $parent->getTranslation($locale)->getSlug().self::SLUG_SEPARATOR.$taxonSlug;
```
Comment (lchrusciel) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php]: Unfortunately, it didn't work. There are two options, we can either move back to your solution or just inject the fallback locale and check if the locale is defined.

Comment (theodormanolescu) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php]: I don't think the issue is here, I think I wrote the Behat test wrong, let me check

Comment (pamil) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php]: Why isn't it working correctly?

File: features/taxonomy/managing_taxons/editing_taxon_slug_in_multiple_locales.feature [hunk 4]
```diff
@@ -32,6 +32,17 @@ Feature: Editing taxon's slug in multiple locales
         And this taxon should have slug "bs" in "Polish (Poland)"
 
     @ui
```
Comment (pamil) [inline features/taxonomy/managing_taxons/editing_taxon_slug_in_multiple_locales.feature]: It should have `@javascript` tag.

File: src/Sylius/Component/Taxonomy/spec/Generator/TaxonSlugGeneratorSpec.php [hunk 5]
```diff
@@ -47,6 +47,19 @@ function it_generates_slug_based_on_new_taxon_name_and_parent_taxon_slug(
         $this->generate('Battle games', 1)->shouldReturn('board-games/battle-games');;
     }
 
+    function it_generates_slug_based_on_new_taxon_name_and_parent_taxon_slug_for_polish_locale(
+        TaxonInterface $parent,
+        TaxonRepositoryInterface $taxonRepository
+    ) {
+        $taxonRepository->find(1)->willReturn($parent);
+        $parent->getSlug()->willReturn('bronie-sredniowieczne');
+        $parent->getTranslation('pl_PL')->willReturn($parent);
```
Comment (pamil) [inline src/Sylius/Component/Taxonomy/spec/Generator/TaxonSlugGeneratorSpec.php]: It does not return itself, it returns `TaxonTranslationInterface`.

File: src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php (hunk 3 above)
Comment (lchrusciel) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php]: I'm afraid it will not resolve `null` locale correctly :(

Comment (pamil) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGenerator.php]: `TranslatableTrait`:
```php
public function getTranslation($locale = null)
{
    $locale = $locale ?: $this->currentLocale;
    if (null === $locale) {
        throw new \RuntimeException('No locale has been set and current locale is undefined.');
    }

    // ...
}
```

Current locale is always set by a listener running after an object is loaded from the ORM.

File: src/Sylius/Component/Taxonomy/spec/Generator/TaxonSlugGeneratorSpec.php (hunk 5 above)
Comment (theodormanolescu) [inline src/Sylius/Component/Taxonomy/spec/Generator/TaxonSlugGeneratorSpec.php]: Updated to `TaxonTranslationInterface`

File: features/taxonomy/managing_taxons/editing_taxon_slug_in_multiple_locales.feature (hunk 4 above)
Comment (theodormanolescu) [inline features/taxonomy/managing_taxons/editing_taxon_slug_in_multiple_locales.feature]: added `@javascript` tag.

File: src/Sylius/Component/Taxonomy/Generator/TaxonSlugGeneratorInterface.php [hunk 6]
```diff
@@ -17,10 +17,11 @@
 interface TaxonSlugGeneratorInterface
 {
     /**
-     * @param string $name
-     * @param int|null $parentId
+     * @param string      $name
+     * @param int|null    $parentId
+     * @param string|null $locale
```
Comment (Zales0123) [inline src/Sylius/Component/Taxonomy/Generator/TaxonSlugGeneratorInterface.php]: Should be reverted.

File: features/taxonomy/managing_taxons/editing_taxon_slug_in_multiple_locales.feature [hunk 7]
```diff
@@ -31,6 +31,17 @@ Feature: Editing taxon's slug in multiple locales
         Then this taxon should have slug "mw" in "English (United States)"
         And this taxon should have slug "bs" in "Polish (Poland)"
 
+    @ui @javascript
+    Scenario: Creating a taxon with an autogenerated slug from parent
+        Given the store has taxonomy named "Medieval weapons" in "English (United States)" locale and "Bronie średniowieczne" in "Polish (Poland)" locale
+        And I want to create a new taxon for "Medieval weapons"
+        When I specify its code as "SIEGE_ENGINES"
+        And I name it "Siege engines" in "English (United States)"
+        And I name it "Silniki oblężnicze" in "Polish (Poland)"
```
Comment (Zales0123) [inline features/taxonomy/managing_taxons/editing_taxon_slug_in_multiple_locales.feature]: I think that `"Machiny oblężnicze"` would be a more appropriate translation ;)

File: src/Sylius/Component/Taxonomy/spec/Generator/TaxonSlugGeneratorSpec.php [hunk 8]
```diff
@@ -39,12 +40,26 @@ function it_implements_taxon_slug_generator_interface()
...
+    function it_generates_slug_based_on_new_taxon_name_and_parent_taxon_slug_for_polish_locale(
+        TaxonInterface $parent,
+        TaxonTranslationInterface $taxonTranslation,
+        TaxonRepositoryInterface $taxonRepository
+    ) {
+        $taxonRepository->find(1)->willReturn($parent);
+        $parent->getTranslation('pl_PL')->willReturn($taxonTranslation);
+        $taxonTranslation->getSlug()->willReturn('bronie-sredniowieczne');
+
+        $this->generate('Silniki oblężnicze', 1, 'pl_PL')->shouldReturn('bronie-sredniowieczne/silniki-obleznicze');
```
Comment (Zales0123) [inline src/Sylius/Component/Taxonomy/spec/Generator/TaxonSlugGeneratorSpec.php]: The same here `Silniki` -> `Machiny` ;)

**Pull request closed without merging**